### DISCUSSION TITLE
husky > Failed to install

### DISCUSSION THREAD

**Discussion by guyellis** [vercel/next.js#16905] [Help]
# Bug report

## Describe the bug

Error `husky > Failed to install` when bootstrapping new project.

## To Reproduce

1. `npx create-next-app --example with-typescript-eslint-jest with-typescript-eslint-jest-app`

## Expected behavior

No husky error

## Screenshots

Output:

```shell
$ npx create-next-app --example with-typescript-eslint-jest with-typescript-eslint-jest-app
npx: installed 1 in 1.037s
Creating a new Next.js app in /home/guy/source/temp/with-typescript-eslint-jest-app.

Downloading files for example with-typescript-eslint-jest. This might take a moment.

Installing packages. This might take a couple of minutes.


> husky@4.3.0 install /home/guy/source/temp/with-typescript-eslint-jest-app/node_modules/husky
> node husky install

husky > Setting up git hooks
fatal: not a git repository (or any of the parent directories): .git
husky > Failed to install

> core-js-pure@3.6.5 postinstall /home/guy/source/temp/with-typescript-eslint-jest-app/node_modules/core-js-pure
> node -e "try{require('./postinstall')}catch(e){}"


> @ampproject/toolbox-optimizer@2.6.0 postinstall /home/guy/source/temp/with-typescript-eslint-jest-app/node_modules/@ampproject/toolbox-optimizer
> node lib/warmup.js

AMP OPTIMIZER Downloaded latest AMP runtime data.

> husky@4.3.0 postinstall /home/guy/source/temp/with-typescript-eslint-jest-app/node_modules/husky
> opencollective-postinstall || exit 0

added 1448 packages from 585 contributors and audited 1452 packages in 23.105s

66 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities


Initialized a git repository.

Success! Created with-typescript-eslint-jest-app at /home/guy/source/temp/with-typescript-eslint-jest-app
Inside that directory, you can run several commands:

  npm run dev
    Starts the development server.

  npm run build
    Builds the app for production.

  npm start
    Runs the built app in production mode.

We suggest that you begin by typing:

  cd with-typescript-eslint-jest-app
  npm run dev
```

## System information

- OS: Ubuntu 20.04
- Browser: N/A
- Version of Next.js: 9.5.3
- Version of Node.js: 14.9.0

**Comment by guyellis** [comment]
My guess as to why this is happening is that `create-next-app` is generating a new folder and bootstrapping that folder and that `husky` (for good reason) needs to be installed while the `pwd` is that folder so that it has access to `.git` for adding the hooks.

For installs where a dependency relies on `.git` being present what do you think of the bootstrap script doing:

- cd <directory>
- install dependency
- cd ..

**Reply by Celtian** [reply]
The same problem with husky.

> 21:36:32.287 | yarn install v1.22.4
-- | --
> 21:36:32.309 | info No lockfile found.
> 21:36:32.314 | [1/4] Resolving packages...
> 21:36:35.755 | warning jest > jest-cli > jest-config > jest-environment-jsdom > jsdom > request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
> 21:36:35.764 | warning jest > jest-cli > jest-config > jest-environment-jsdom > jsdom > request-promise-native@1.0.9: request-promise-native has been deprecated because it extends the now deprecated request package, see https://github.com/request/request/issues/3142
> 21:36:35.907 | warning jest > jest-cli > jest-config > jest-environment-jsdom > jsdom > request > har-validator@5.1.5: this library is no longer supported
> 21:36:36.044 | warning jest > @jest/core > jest-haste-map > sane > micromatch > snapdragon > source-map-resolve > resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
> 21:36:36.064 | warning jest > @jest/core > jest-haste-map > sane > micromatch > snapdragon > source-map-resolve > urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
> 21:36:36.546 | [2/4] Fetching packages...
> 21:36:45.503 | info fsevents@2.1.3: The platform "linux" is incompatible with this module.
> 21:36:45.504 | info "fsevents@2.1.3" is an optional dependency and failed compatibility check. Excluding it from installation.
> 21:36:45.511 | info fsevents@2.2.1: The platform "linux" is incompatible with this module.
> 21:36:45.511 | info "fsevents@2.2.1" is an optional dependency and failed compatibility check. Excluding it from installation.
> 21:36:45.517 | [3/4] Linking dependencies...
> 21:36:45.523 | warning " > ts-loader@8.0.11" has unmet peer dependency "webpack@*".
> 21:36:52.289 | [4/4] Building fresh packages...
> 21:36:52.436 | error /vercel/workpath0/node_modules/husky: Command failed.
> 21:36:52.436 | Exit code: 1
> 21:36:52.436 | Command: node husky install
> 21:36:52.436 | Arguments:
> 21:36:52.436 | Directory: /vercel/workpath0/node_modules/husky
> 21:36:52.436 | Output:
> 21:36:52.437 | husky > Setting up git hooks
> 21:36:52.437 | fatal: not a git repository (or any of the parent directories): .git
> 21:36:52.437 | husky > Failed to install
> 21:36:52.437 | info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
> 21:36:52.476 | Error: Command "yarn install" exited with 1

Workaround is to move husky into optionalDependencies and do not install them in vercel.

**Reply by peiris** [reply]
@Celtian Thank you. Moving husky to optionalDependencies worked for me. You saved me 😄

"optionalDependencies": {
    "husky": "^4.3.4"
  }

**Reply by peiris** [reply]
PS: This is the same case with create-react-app as well. not just next.

**Comment by adamsir** [comment]
Possible workaround is to prefix the husky command  with `npx` like:
```
"scripts": {
  "prepare": "npx husky install",
}
```

**Reply by gaspardip** [reply]
> Possible workaround is to prefix the husky command with `npx` like:
> 
> ```
> "scripts": {
>   "prepare": "npx husky install",
> }
> ```

Worked for me.

**Reply by isaacdarcilla** [reply]
Worked for me, in my case, it was in a custom directory.

```json
"prepare": "cd ../../ && npx husky install"
```

**Reply by adamsir** [reply]
@guyellis check this out

**Comment by guyellis** [comment]
Thanks @adamsir - I can't even remember what project of mine this was happening on. I'm going to close this issue but feel free to open it again if someone else wants to continue this. Thanks for your help.
